tripal/tripal
# Tripal 4 add a chado boolean field

# Tripal 4 Core Dev Task 
# Add a chado boolean field

## Issue #1551 

### Tripal Version: 4.alpha1-dev

## Description
We are missing a boolean field for use on e.g. `is_obsolete` for the feature table, such as a gene, stock, and pub content types, and others.
Note that some `is_obsolete` columns are integers rather than boolean, such as for the library table. This field will not be appropriate for those.
The yaml for pub is currently not present in `tripal_chado/config/install/tripal.tripalfield_collection.default_chado.yml`, so that field will not be added in this pull request.
A few related typos and enhanced error messages (that I triggered while developing this) are in here also.

## Testing?
I have added boolean fields for all content types that have them as boolean (as opposed to integer), see the table below. I had to add an organism field to most of the feature content types, only gene had this field, so that we can test those also. Other missing fields will be added with issue #1550.
 1. Create a docker from this branch (see instructions [here](https://tripaldoc.readthedocs.io/en/latest/contributing/docker-for-testing.html).)
 2. Create an organism (needed for next steps).
 3. Create a new gene, enter a zero for sequence length (that's issue #1387), and do not check "Is Analysis" or "Is Obsolete". Save and verify it displays "False" for both.
 4. Edit the gene, verify the check boxes are unchecked, then check both, and save. Verify it now says "True" for both.
 5. Create a second gene and check "Is Analysis" and "Is Obsolete" when creating. Save and verify both display "True".
 6. Edit the gene, verify the check boxes are both checked, then uncheck both, and save. Verify it now says "False" for both.
 7. Test similarly on mrna, qtl, sequence_variant, genetic_marker, phenotypic_marker
 8. Test similarly on the stock content types, these only have a single "Is Obsolete" field: germplasm, breeding_cross, germplasm_variety, ril
 9. Pub is not implemented here because it is entirely missing from the yaml so that is beyond the scope of this pull request.

## List of boolean fields in Chado 1.3 tables

Only the tables in bold are content types

| boolean field and default | tables |
| ------------------------------------ | --------- |
| is_analysis boolean not null default 'false' | **feature** |
| is_current boolean not null default 'false' | feature_synonym, cell_line_synonym |
| is_current boolean not null default 'true' | pub_dbxref, feature_dbxref, analysis_dbxref, featuremap_dbxref, library_synonym, library_dbxref, stock_dbxref, project_dbxref, cell_line_dbxref |
| is_fmax_partial boolean not null default 'false' | featureloc |
| is_fmin_partial boolean not null default 'false' | featureloc |
| is_internal boolean not null default 'false' | feature_synonym, library_synonym, cell_line_synonym |
| is_not boolean not null default false | feature_cvterm, analysis_cvterm, stock_cvterm |
| is_obsolete boolean default 'false' | **pub** |
| is_obsolete boolean not null default 'false' | **feature**, **stock** |

Content types using **feature** table (6): gene, mrna, qtl, sequence_variant, genetic_marker, phenotypic_marker
Content types using **stock** table (4): germplasm, breeding_cross, germplasm_variety, ril
Content types using **pub** table (1): publication
